### PR TITLE
support as extendible and isextendible

### PR DESCRIPTION
**Pull request by huberya**
Is this more or less what you had in mind?

### CODE REVIEW

**Comment by HannesSommer** [comment]
Thanks already a. It is not quite - but good direction :). Sorry, if I wasn't clear enough. 
You only need these two functions on the `ValueStoreRef` (non virtual as all the *Ref methods should be) and on the `ValueStore` interface itself. Then the different implementations you have are for the classes without `Ref` suffix. Otherwise it wont work. Maye add a minor unit test to see for yourself.

**Comment by huberya** [comment]
I'm a bit confused:
I wrote a unit test (which works)
1. I dont understand why it shouldnt work
2. I assume the idea of non virtual in ValueStoreRef would be something like
```
bool isExtendible() {
return vs_->isExtendible();
}
```
and then implement true/false for ValueStore and ExtendibleValueStore

but `vs_` is always non extendible, ExtendibleValueStoreRef has an additional member `evs_` so I dont understand how that would work

**Comment by huberya** [comment]
ah nvm I'm not even using virtual inheritance in my test...

**Comment by huberya** [comment]
now the test fails (as it should)
I still dont understand why and how to fix it though

**Comment by HannesSommer** [comment]
Now I'm confused. If vs_ is always non-extendible (as runtime type not declared type) then this is a bug IMO... Looking into the current code it looks indeed as if sometimes vs_ is null instead of always pointing to the same object as `evs_`. But this can easily be corrected by delegating to the second constructor of the EVS with all the ones below.  The fact that there is another `evs_` having the same value as the `vs_` was simply a very hacky and lacy work around proper static pointer casting while keeping the guarantee programmatically - something that should be cleaned up at some point.

In any case. Once vs_ and evs_ always point to the same object, my suggested way works with this modification (I didn't think enough): Only the `isExtendible` should be delegating to the `vs_` pointer, just as you ask above. The `asExtendible` must be implemented as a normal function of ValueStoreRef doing a static_pointer_cast into an ExtendibleValueStore after checking the isExtendible and feeding the EVS point into a new EVSRef object that it returns by value. Generally think of the *Ref classes as a self made smart ref (i.e. something like a smart pointer that can be cased if the runtime type of the object pointed as allows it).

**Comment by huberya** [comment]
Actually I realised that its always extendible...
Both refs use a PropertyTreeValueStore as the actual object, which is a child of ExtendibleValueStore
I guess thats a bug already?

**Review by HannesSommer — COMMENTED** [review]
Oh my, apparently I never sent these comments ARRRRG. Sorry, that wasted probably some of your time. Can you take them and rethink the changes you made? In particular I don't think we need the two version of PropertyTreeValueStore. IMO it is perfectly fine that this one is always extendible.

**Comment by huberya** [comment]
I think now it should be fine @HannesSommer 
Can you TAL and approve if thats ok
(always extendible with a class for unittest defined locally)

File: sm_value_store/test/ValueStore.cpp [hunk 1]
```diff
@@ -19,20 +19,28 @@ TEST(ValueStoreSuite, isEmpty)
...
 TEST(ValueStoreSuite, isExtendible)
 {
   sm::BoostPropertyTree pt;
-  sm::BoostPropertyTree pt2;
-
-  sm::ValueStoreRef vpt(pt);
-  sm::ValueStoreRef evpt = sm::ExtendibleValueStoreRef(pt2);
+  sm::NonExtendiblePropertyTreeValueStore pt2 = sm::NonExtendiblePropertyTreeValueStore(pt);
+  sm::ValueStore& vs = pt2;
+  sm::ValueStoreRef vpt(vs);
```
Comment (HannesSommer) [inline sm_value_store/test/ValueStore.cpp:36]:
```suggestion
  sm::ValueStoreRef vpt(pt2);
```

**Review by HannesSommer — APPROVED** [review]
Great! Thanks!

File: sm_value_store/test/ValueStore.cpp [hunk 2]
```diff
@@ -19,6 +19,29 @@ TEST(ValueStoreSuite, isEmpty)
...
+    return false;
+  }
+};
+}  // namespace sm
+
+TEST(ValueStoreSuite, isExtendible)
+{
+  sm::BoostPropertyTree pt;
+  sm::NonExtendiblePropertyTreeValueStore pt2 = sm::NonExtendiblePropertyTreeValueStore(pt);
+  sm::ValueStoreRef vpt(pt2);
```
Comment (HannesSommer) [inline sm_value_store/test/ValueStore.cpp:35]:
```suggestion
  sm::ValueStoreRef vpt(static_cast<sm::ValueStore&>(pt2));
```

**Pull request merged**